text/0000-monorepo-support.md

```diff
@@ -0,0 +1,81 @@
+- The behavior of `volta pin` in subprojects vs the root project.
+
+Ideally, we should document all of these with specific examples to show how the different pieces interact. We could potentially even create an `examples` repo (or directory within the main repo) for users to reference.
+
+# Details
+[details]: #details
+
+## Multiple Roots
+
+The core change to Volta will be expanding our `Project` model to support a _sequence_ of project roots, instead of a single project root. We will support a sequence by adding the ability to put an `"extends"` entry inside the `"volta"` section of `package.json`. The value of that entry will be the path to another json file, which we will parse just like the original `package.json`. The directory that new file is in will be the next value in the sequence of project roots. Then, if that new file also has an `"extends"` entry, we will continue the same process, building up effectively a linked list of project roots to use for all of our project information.
+
+Importantly, the order of the sequence of project roots will start with the closest `package.json` (what we previously called the sole project root), and then follow the `"extends"` chain out from there.
+
+## Platform Resolution
+
+To resolve the Project platform, taking into account all of the settings in the different project files, we will merge the platform settings from each project file, giving precedence at all times to the files earlier in the sequence. So the highest precedence will be at the existing project root, and then each file that we resolve after that will only update settings that haven't been specified by an earlier entry.
+
+To support a use-case where a project wants to exclude e.g. `yarn` and _not_ allow a later project file to add one, we will support a new value for the package manager keys in the `"volta"` section: `false`. Setting `"yarn": false` will mean that no Yarn is available _and_ that a file later in the sequence with a `yarn` setting won't be able to override that removal of Yarn.
+
+Additionally, while we are processing this chain of project files, we will need to take care to detect accidental loops in the files pointed to by `"extends"` and exit early with a helpful error message (as opposed to entering an infinite loop of parsing and never finishing).
+
+Finally, in lieu of the current parse error for not setting `"node"` in the `"volta"` section, we should merge all of the project settings into a single project Platform, and then only show an error if none of the files specified a Node version.
+
+## Local Binary Resolution
+
+Currently, in order to run a project-local binary, we look for the executable in `<project root>/node_modules/.bin/`. With multiple project roots, this process will be almost the same: For each root in the sequence, we check for the binary in the expected place, and if it's found, then we stop and use that version. If we don't find it, we move on to the next root in the sequence.
```

Comment:
I think this section would also be much easier to understand with a small diagram. I'm thinking something like:

---

Given the following folder structure:

```
.
├── package.json
└── packages
   ├── bar
   │  └── package.json
   └── foo
      └── package.json
```

When a previously `volta install`ed binary is executed from within the `packages/foo/` subdirectory we will look for a local installation of that binary in the following locations:

1. `packages/foo/node_modules/.bin/<binary name>
2. `node_modules/.bin/<binary name>
3. Fall back to our own global (sorry, I know that isn't the specific term we use, but I couldn't remember the right term)

Reply:
Definitely, I like the tree layout to help make a compact example!